### PR TITLE
Two-staged RecSync workflow

### PR DESCRIPTION
**Pull request by TimPushkin**
RecSync workflow is changed with recording start divided into two stages:
1. Video preparation (up to and including the `camera_controller.initVideoRecorderPostPrepare` call)
2. Recording start (after the `camera_controller.initVideoRecorderPostPrepare` call)

Related fixed bugs:
- When a client connects to a leader which has settings sync enabled, the client either gets an app crash ar a camera error.
  - Solution: video recording preparation was programmed to start each time the camera reopens but only if `SoftwareSyncController`'s field `mIsVideoPreparationNeeded` equals `false` -- the problems was that this field was set to `true` before settings sync (which might reopen the camera) finishes, so the preparation was starting before settings sync finishes and it led to crashes.
- When settings sync is started the second time it never finishes.
  - Solution: `applyAndLockSettings()` has been incorrectly determining the need for a camera reopening.
- Settings sync never truly finishes (i.e. with the `onFinished()` invocation being finished) when started from the "Sync settings" button click.
  - Solution: `Preview,prepareVideoRecording()` call needs to be made inside a `MainActivity.runOnUiThread()` block, or otherwise call that need a `Looper` (such as [this one](https://github.com/MobileRoboticsSkoltech/OpenCamera-Sensors/blob/3122e4f97e046924839956eb3303c9993c0201cd/app/src/main/java/net/sourceforge/opencamera/cameracontroller/CameraController2.java#L7241)) cannot be executed.
- Clients sometimes have camera errors when settings sync is cancelled (specifically, when the recording preparation is cancelled).
  - Solution: `Preview.stopVideo()` call needs to be made inside a `MainActivity.runOnUiThread()` block (with the same reasoning as above).

### CODE REVIEW

File: app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java [hunk 1]
```diff
@@ -204,10 +210,25 @@ private void setupSoftwareSync() {
 
                     if (settings != null) {
                         mState = State.SETTINGS_APPLICATION;
-                        mMainActivity.getApplicationInterface().getSoftwareSyncUtils().applyAndLockSettings(settings, () -> mState = State.IDLE);
+                        mMainActivity.getApplicationInterface().getSoftwareSyncUtils().applyAndLockSettings(settings, () -> {
+                            mMainActivity.getApplicationInterface().getSoftwareSyncUtils().prepareVideoRecording();
```
Comment (DmiitriyJarosh) [inline app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java:214]: Extract softwareSyncUtils to variable

File: app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java [hunk 2]
```diff
@@ -249,21 +280,26 @@ private void setupSoftwareSync() {
...
+            // Adjust the state when establishing a connection to a leader.
             clientRpcs.put(
                     SyncConstants.METHOD_MSG_SYNCING,
-                    payload ->
-                            mMainActivity.runOnUiThread(
-                                    () -> mSyncStatus = mMainActivity.getString(R.string.rec_sync_waiting_for_sync, mSoftwareSync.getName())));
+                    payload -> {
+                        mIsVideoPreparationNeeded = false;
+                        mMainActivity.runOnUiThread(
+                                () -> mSyncStatus = mMainActivity.getString(R.string.rec_sync_waiting_for_sync, mSoftwareSync.getName()));
```
Comment (DmiitriyJarosh) [inline app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java:299]: Is it possible now to fully remove links to mainActivity from this controller?

File: app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java [hunk 3]
```diff
@@ -302,20 +338,25 @@ private void setupSoftwareSync() {
         @RequiresApi(api = Build.VERSION_CODES.N)
         @Override
         protected Void doInBackground(Void... voids) {
+            Log.d(TAG, "Starting AlignPhasesTask");
```
Comment (DmiitriyJarosh) [inline app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java:341]: Lets extract this asynctask to separate file

File: app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java [hunk 4]
```diff
@@ -430,6 +478,15 @@ public boolean isSettingsBroadcasting() {
         throw new IllegalStateException("Cannot check the settings lock status for a client");
     }
 
+    /**
+     * Shows whether is device is expected to be prepared for video recording.
```
Comment (DmiitriyJarosh) [inline app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java:482]: whether THE device?

File: app/src/main/java/net/sourceforge/opencamera/preview/Preview.java [hunk 5]
```diff
@@ -5765,6 +5722,104 @@ protected void onPostExecute(Boolean success) {
...
+            // needed for emulator at least - although MediaRecorder not meant to work with emulator, it's good to fail gracefully
+            Log.e(TAG, "runtime exception starting video recorder");
+            e.printStackTrace();
+            this.video_recorder = local_video_recorder; // still assign, so failedToStartVideoRecorder() will release the video_recorder
+            // told_app_starting must be true if we're here
+            applicationInterface.stoppingVideo();
+            failedToStartVideoRecorder(profile);
+        }
+
+        /*final MediaRecorder local_video_recorder_f = local_video_recorder;
```
Comment (DmiitriyJarosh) [inline app/src/main/java/net/sourceforge/opencamera/preview/Preview.java:5785]: A lot of commented code in this file --- why we still need it?

**Review by DmiitriyJarosh — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java (hunk 2 above)
Comment (TimPushkin) [reply app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java:299]: I don't think so as it is currently heavily used all over the controller. I can replace calls like to `getPreview()` and `getExtendedAppInterface()` with the corresponding fields, but the controller still will use `getString()` from `MainActivity` as a `Context`, `runOnUiThread()` from it as an `Activity`, `getResources()`, etc.

File: app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java (hunk 3 above)
Comment (TimPushkin) [reply app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java:341]: It works with `SoftwareSyncController`'s private fields such as `mState` and `mIsPeriodCalculated`, which, as I think, is not a good idea to expose.

I also noticed that `AsyncTask` became deprecated in API 30.

File: app/src/main/java/net/sourceforge/opencamera/preview/Preview.java (hunk 5 above)
Comment (TimPushkin) [reply app/src/main/java/net/sourceforge/opencamera/preview/Preview.java:5785]: This huge piece was actually left there by the original developers so I left it untouched too. I will remove the little pieces above created by me as they are not needed anymore, but I'm not sure what to do with this one.

Comment (TimPushkin) [reply app/src/main/java/net/sourceforge/opencamera/preview/Preview.java:5785]: This huge piece was actually left there by the original developers so I left it untouched too. I will remove the little pieces above created by me as they are not needed anymore, but I'm not sure what to do with this one.

File: app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java (hunk 3 above)
Comment (DmiitriyJarosh) [reply app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java:343]: Ok, I agree

File: app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java (hunk 2 above)
Comment (DmiitriyJarosh) [reply app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java:301]: All can be changed, for example, strings need only appcontext, UI thread launch can be done without activity at all (via handlers), but lets skip this now

File: app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java (hunk 1 above)
Comment (DmiitriyJarosh) [reply app/src/main/java/com/googleresearch/capturesync/SoftwareSyncController.java]: I think that there are some more places in file with usage of it, not only one string to fix, lets change every where

**Review by DmiitriyJarosh — APPROVED** [review]

**Pull request merged**